File: 2021/Python/day-02.py
import pandas as pd
import unittest


def day_02a(DataFrame) :
    x = DataFrame.groupby("direction").sum()
    res = x["value"]["forward"] * (x["value"]["down"] - x["value"]["up"])
    return res


def day_02b(DataFrame) :
  # initialize values at 0
  aim = 0
  horizontal = 0
  depth = 0
  
  for i in range(len(DataFrame)) :
      step = DataFrame["direction"][i]
      
      if step == "down" :
          aim = aim + DataFrame["value"][i]
      elif step == "up" :
          aim = aim - DataFrame["value"][i]
      elif step == "forward" :
          horizontal = horizontal + DataFrame["value"][i]
          depth = depth + aim * DataFrame["value"][i]
      else :
          raise ValueError(f'"{step}" is not a valid step')
  
  res = horizontal * depth
  return res


class testDay02(unittest.TestCase) :

    def test_day_02a(self) :
        self.assertEqual(day_02a(df), 1459206)
  
    def test_day_02b(self) :
        self.assertEqual(day_02b(df), 1320534480)
  

if __name__ == '__main__':
    df = pd.read_table("data/day-02.txt", delimiter=" ", names=["direction", "value"])
    unittest.main()

File: 2021/R/day-02.R

x <- readLines("data/day-02.txt")
x <- strsplit(x, "\\s")
s <- seq_along(x)
steps <- vapply(s, \(i) as.integer(x[[i]][2]), NA_integer_)
names(steps) <- vapply(s, \(i) x[[i]][1], NA_character_)
sums <- tapply(steps, names(steps), sum)
res <- (sums[["down"]] - sums[["up"]]) * sums[["forward"]]

stopifnot(identical(res, 1459206L))

aim <- 0L
horizontal <- 0L
depth <- 0L

for (i in s) {
  switch(
    names(steps[i]),
    down = {
      aim <- aim + steps[[i]]
    },
    up = {
      aim <- aim - steps[[i]]
    },
    forward = {
      horizontal <- horizontal + steps[[i]]
      depth <- depth + aim * steps[[i]]
    }
  )
}

res <- horizontal * depth
stopifnot(identical(res, 1320534480L))
